### ISSUE TITLE
Maybe our DCGAN have extra nn.BatchNorm2d(ndf)?

### ISSUE THREAD

**Issue by StevenJokes** [d2l-ai/d2l-en#1381]
First, I think pytorch tutorial is wrong. "DCGAN's Discriminator : lack of nn.BatchNorm2d(ndf) "

https://github.com/pytorch/tutorials/issues/1126

![image](https://user-images.githubusercontent.com/25657787/90494474-1e5ca500-e176-11ea-911f-3cf8207aea21.png)

>I'm a newbie. Then in my guess, Discriminator has pretty much structures opposite to Generator. So they must have same number layers.
>Here is my code:
>https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/dcgan_pt3.ipynb
>My PR: #1125

Then I went to https://github.com/soumith/dcgan.torch/blob/master/main.lua, and found no SpatialBatchNormalization(ndf) neither
![image](https://user-images.githubusercontent.com/25657787/90494205-d3429200-e175-11ea-97e6-5753b982fb78.png)

Did we use `D_block(n_D)` abusively?

**Comment by StevenJokes** [comment]
https://www.zhihu.com/pub/reader/119607400/chapter/1182323947911188480
Also no nn.BatchNorm2d(ndf) 
![image](https://user-images.githubusercontent.com/25657787/90641297-a79dd580-e263-11ea-8a38-8f41d5587521.png)

I also asked 深度学习之模型设计：核心算法与案例实践 book's author 言有三. His tensorflow code didn't have this layer too.

**Comment by StevenJokes** [comment]
@mli

**Comment by soumith** [comment]
hey, BatchNorm seems unnecessary for the first layer, as the inputs to the first Conv2d are already zero-mean and unit std, and the Conv2d is initialized (initially in a random draw) in a way that the output of the Conv2d roughly maintains zero-mean and unit std.

This is why we avoided putting batchnorm in the first layer in our original code / paper

**Comment by StevenJokes** [comment]
Thanks for your reply.@soumith
Unnecessary means that it is OK to add the layer if we use D_block way to simplify code?

**Comment by StevenJokes** [comment]
And BTW, in my stupid code, I found pytorch is slower than mxnet. Can you help me improve the pytorch code...

**Comment by soumith** [comment]
> Unnecessary means that it is OK to add the layer if we use D_block way to simplify code?

Yes you can add if you want.

> And BTW, in my stupid code, I found pytorch is slower than mxnet. Can you help me improve the pytorch code...

Maybe PyTorch is slower than MXNet :)

**Comment by StevenJokes** [comment]
@mli 
Would you like to give me an offer?
Almost crazy about staying home with no income...

**Comment by StevenJokes** [comment]
Die DIe DIe DIE aaaaa Fk. I hate world. Non-sense. Shit！！！！
You all go die...
